vwifi.c

```diff
@@ -846,17 +846,102 @@ static int owl_delete_interface(struct owl_vif *vif)
     return 0;
 }
 
+/* Called by kernel when user decided to change the interface type. */
+static int owl_change_iface(struct wiphy *wiphy,
+                            struct net_device *ndev,
+                            enum nl80211_iftype type,
+                            struct vif_params *params)
+{
+    switch (type) {
+    case NL80211_IFTYPE_STATION:
+    case NL80211_IFTYPE_AP:
+        ndev->ieee80211_ptr->iftype = type;
+        break;
+    default:
+        pr_info("owl: invalid interface type %u\n", type);
+        return -EINVAL;
+    }
+
+    return 0;
+}
+
+/* Called by the kernel when user want to create an Access Point. Now
+ * it just add a ssid to the ssid_table to emulate the AP signal. And
+ * record the ssid to the owl_context.
+ */
+static int owl_start_ap(struct wiphy *wiphy,
+                        struct net_device *ndev,
+                        struct cfg80211_ap_settings *settings)
+{
+    struct owl_vif *owl = ndev_get_owl_vif(ndev);
+
+    pr_info("owl: start acting in AP mode.\n");
+    pr_info("ctrlchn=%d, center=%d, bw=%d, beacon_interval=%d, dtim_period=%d,",
+            settings->chandef.chan->hw_value, settings->chandef.center_freq1,
+            settings->chandef.width, settings->beacon_interval,
+            settings->dtim_period);
+    pr_info("ssid=%s(%zu), auth_type=%d, inactivity_timeout=%d", settings->ssid,
+            settings->ssid_len, settings->auth_type,
+            settings->inactivity_timeout);
+
+    if (settings->ssid == NULL)
+        return -EINVAL;
+
+    /* set AP SSID */
+    memcpy(owl->ssid, settings->ssid, settings->ssid_len);
+    update_ssids(owl->ssid);
+
+    return 0;
+}
+
+/* Called by the kernel when there is need to "stop" from AP mode.
+ * It uses the ssid to remove the AP node of ssid_stable.
+ */
+static int owl_stop_ap(struct wiphy *wiphy, struct net_device *ndev)
+{
+    struct owl_vif *item, *vif = ndev_get_owl_vif(ndev);
+
+    struct ap_info_entry_t *ap;
+    struct hlist_node *tmp;
+    int i;
+
+    pr_info("owl: stop acting in AP mode.\n");
+    if (hash_empty(ssid_table))
+        return -EINVAL;
+
+    hash_for_each_safe (ssid_table, i, tmp, ap, node) {
+        if (!strncmp(ap->bssid, vif->bssid, ETH_ALEN)) {
+            /* Remove AP from ssid_table by bssid */
+            hash_del(&ap->node);
+
+            list_for_each_entry (item, &owl->vif_list, list) {
+                if (!strncmp(vif->bssid, item->bssid, ETH_ALEN) &&
+                    item->sme_state == SME_CONNECTED)
+                    /* Disconnect STA if some STA connected AP */
```

Comment:
See if we shall use intransitive verb for connect.
Reference: https://www.merriam-webster.com/dictionary/connect

Reply:
ok